src/permissionSelectors.js

```diff
@@ -1,9 +1,48 @@
 /* @flow strict-local */
 import type { PerAccountState } from './types';
 import { ensureUnreachable } from './types';
-import { Role } from './api/permissionsTypes';
+import { Role, type RoleT } from './api/permissionsTypes';
 import { getRealm, getOwnUser } from './selectors';
 
+const { Guest, Member, Moderator, Admin, Owner } = Role;
+
+/**
+ * The Role of the self-user.
+ *
+ * For servers at FL 59 (version 4.0) or above, this will be live-updated
+ * when the role changes while we're polling an event queue.
+ */
+// TODO(server-4.0): Probably just delete this and use User.role directly?
+export function getOwnUserRole(state: PerAccountState): RoleT {
+  const fromUserObject = getOwnUser(state).role;
+
+  if (fromUserObject !== undefined) {
+    return fromUserObject;
+  }
+
+  // Servers don't send events to update these.
+  const { isOwner, isAdmin, isModerator, isGuest } = getRealm(state);
+
+  if (isOwner) {
+    return Owner;
+  } else if (isAdmin) {
+    return Admin;
+  } else if (isModerator) {
+    return Moderator;
+  } else if (isGuest) {
+    return Guest;
+  } else {
+    return Member;
+  }
+}
+
+// PRIVATE: exported only for tests.
+export const kRolesAscending = [Guest, Member, Moderator, Admin, Owner];
+
+export function roleIsAtLeast(thisRole: RoleT, thresholdRole: RoleT): boolean {
+  return kRolesAscending.indexOf(thisRole) >= kRolesAscending.indexOf(thresholdRole);
```

Comment:
I'd be a bit concerned about the efficiency of this implementation -- it goes and searches through an array twice. This check is something that fundamentally ought to be constant-time and cheap.

Because it fundamentally ought to be cheap, if there's some situation where we want the result in a loop (perhaps for some bit of UI on each of a list of users or streams?), we might naturally just call it repeatedly, without attempting some optimization like caching it.

Here's a version that would work, in constant time:
```js
export function roleIsAtLeast(thisRole: RoleT, thresholdRole: RoleT): boolean {
  switch (thresholdRole) {
    // FlowIssue: sad numeric literals here; see similar comment below
    case 100: // Owner
      return thisRole === Owner;
    case 200: // Admin
      return thisRole === Owner || thisRole === Admin;
    case 300: // Moderator
      return thisRole === Owner || thisRole === Admin || thisRole === Moderator;
    case 400: // Member
      return !(thisRole === Guest);
    case 600: // Guest
      return true;
    default: {
      ensureUnreachable(thresholdRole);
      return false;
    }
  }
```

It's more code; but it's all boring straightforward code.

Then the list in `kRolesAscending`, and this `indexOf` implementation, is good for a set of unit tests -- an inefficient, but short and clear, implementation is great for that. So the tests can basically stay just like in this version, but with `kRolesAscending` as a local instead of importing it from here.